### DISCUSSION TITLE
useState persisted?

### DISCUSSION THREAD

**Discussion by Coder-Iro** [nuxt/nuxt#16067] [Questions]
There is no vuex in nuxt 3. I tried to create persisted state using https://github.com/posva/pinia and https://github.com/prazdevs/pinia-plugin-persistedstate. But it didn't work. Is there any way to create persisted state using nuxt's internal state?

**Comment by wobsoriano** [comment]
Can you show a simple repro of what you've done so far?

Have you tried [my plugin](https://github.com/wobsoriano/pinia-shared-state) too? 

I think using nuxt's internal state, you can watch it and set localstorage

```ts
// composables/states.ts
export const useCounter = () => useState<number>('counter', () => {
    return localStorage.getItem('STATE_COUNT') ?? 0
})
```

```html
<script setup lang="ts">
const counter = useCounter()

watch(counter, (count) => {
    localStorage.setItem('STATE_COUNT', count)
})
</script>

<template>
  <div>
    Some shared layout content:
    <slot />
  </div>
</template>
```

```html
<script setup lang="ts">
const counter = useCounter()
</script>

<template>
  <p>Current count: {{ counter }}</p>
</template>
```

**Reply by Coder-Iro** [reply]
I made it after uploading this discussion. [Link](https://github.com/Coder-Iro/wonderland-homepage/blob/dev/plugins/useUser.ts) However, cookies, not localstorage, must be used to use the store in universal mode. 
 And in order to use cookies, the store must be defined within the plugin. I really hate this.

**Reply by pi0** [reply]
I wish `pessist.storage` could allowed to be a function and lazy evaluated within pinia's own plugin. This way we could use `useCookie` composable to sync /cc @posva

**Reply by posva** [reply]
I don't know what `persist.storage` is but  can't you use `useCookie()` inside a pinia plugin? Plugins must be sync because they set properties on the store

**Reply by Coder-Iro** [reply]
I tryed to define store in composable, but I cannot call useCookie in composable
```ts
import { defineStore } from 'pinia';

const user = useCookie('user', { expires: 365, secure: false });

const useUser = defineStore({
  id: 'user',
  state: () => ({
    username: '',
    uuid: '',
    token: '',
    isLoggedIn: false,
  }),
  actions: {
    login({
      username,
      uuid,
      token,
    }: {
      username: string;
      uuid: string;
      token: string;
    }) {
      this.isLoggedIn = true;
      this.username = username;
      this.uuid = uuid;
      this.token = token;
    },
    logout() {
      this.isLoggedIn = false;
      this.username = '';
      this.uuid = '';
      this.token = '';
    },
  },
  persist: {
    storage: {
      getItem: (key: string) => user.value[key] || null,
      setItem: (key: string, value: string) => {
        user.value[key] = value;
      },
      removeItem: (key: string) => {
        delete user.value[key];
      },
    },
  },
});

export default useUser;
```

```
nuxt instance unavailable
  at Module.useNuxtApp (file://./.nuxt/dist/server/server.mjs:275:13)  
  at useSSRReq (file://./.nuxt/dist/server/server.mjs:799:52)  
  at readRawCookies (file://./.nuxt/dist/server/server.mjs:807:40)  
  at Module.useCookie (file://./.nuxt/dist/server/server.mjs:781:19)  
  at $id_22e311b4 (file://./.nuxt/dist/server/server.mjs:934:36)  
  at async __instantiateModule__ (file://./.nuxt/dist/server/server.mjs:4750:3)
```

**Reply by eliasmikkola** [reply]
Did you come up with any sorts of workaround for this? I'd also like to set cookies on Pinia store's action (the action might be called during SSR).

**Reply by itzaks** [reply]
I wish nuxt supported something like 
````ts 
useState<string>('key', () => { 'defaultValue' }, { storage: 'local' })
```` 

using something like Vue's useLocalStorage under the hood for saving the data.

For the moment I'm doing a workaround like this:

````ts
import type { Ref } from 'vue'

export const usePersistedState = <T extends Object>(identifier: string, defaultOptions: T): Ref<T> => {
  const persistedObject = useState<T>(identifier, (): T => {
    const item = localStorage.getItem(identifier)
    if (!item)
      return defaultOptions

    return JSON.parse(item) as T ?? defaultOptions
  })

  watch(persistedObject, (object) => {
    localStorage.setItem(identifier, JSON.stringify(object))
  }, { deep: true })

  return persistedObject
}
````

Used like this:
````ts
 const user = usePersistedState('user', {})
````

**Reply by mohsin** [reply]
Yep, totally this would be excellent. The third parameter can be a configurable adapter i.e. either store in predefined storages such as 'cookie', 'local' or even user defined ones such as passing it thro' an API call to store in server or whatever other logic.

**Reply by notflip** [reply]
> Can you show a simple repro of what you've done so far?
> 
> Have you tried [my plugin](https://github.com/wobsoriano/pinia-shared-state) too?
> 
> I think using nuxt's internal state, you can watch it and set localstorage
> 
> ```ts
> // composables/states.ts
> export const useCounter = () => useState<number>('counter', () => {
>     return localStorage.getItem('STATE_COUNT') ?? 0
> })
> ```
> 
> ```
> <script setup lang="ts">
> const counter = useCounter()
> 
> watch(counter, (count) => {
>     localStorage.setItem('STATE_COUNT', count)
> })
> </script>
> 
> <template>
>   <div>
>     Some shared layout content:
>     <slot />
>   </div>
> </template>
> ```
> 
> ```
> <script setup lang="ts">
> const counter = useCounter()
> </script>
> 
> <template>
>   <p>Current count: {{ counter }}</p>
> </template>
> ```

Thanks for this example, however I'm receiving an error "localStorage" is not defined, since I'm using SSR I suppose the server is trying to use localStorage, which of course is not possible.

**Comment by muellerluke** [comment]
Anything new on this? Are there any new solutions?

**Reply by Coder-Iro** [reply]
useCookie is itself state. You can just use useCookie composable.
